gh-134309: Fix premature cancellation of CI tests when multiple PRs have the same branch name

abstractedfox opened the pull request:
Per [GitHub](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#example-concurrency-groups-1), "When you define a concurrency key, GitHub Actions ensures that only one workflow or job with that key runs at any given time." Under the current config, concurrency group names for pull requests are made unique (against those of other pull requests) only by the name of the branch, so if pull requests were made for identically named branches around the same time, CI would fail. (This affected commit #134234 )

This PR addresses that by using more unique attributes to name each concurrency group. This should also help with debugging any future issues that concern the name of a concurrency group.

<!-- gh-issue-number: gh-134309 -->
* Issue: gh-134309
<!-- /gh-issue-number -->

abstractedfox committed:
Add tests for zero-sized bytes objects in test_bytes.py

abstractedfox committed:
Merge branch 'main' into main

abstractedfox committed:
Merge branch 'main' into main

abstractedfox committed:
Fix trailing whitespace

abstractedfox committed:
Fix trailing whitespace

abstractedfox committed:
Merge branch 'main' into main

abstractedfox committed:
Merge branch 'main' into main

abstractedfox committed:
Reorder lines

abstractedfox committed:
Fix CI jobs with the same concurrency group being canceled

Addresses issue encountered during PyCon US sprints where having multiple workflows with the same concurrency group causes one of the two to be canceled.

Co-authored-by: Sviatoslav Sydorenko <sviat@redhat.com>
Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

abstractedfox committed:
Revert changes made on the wrong branch

AA-Turner commented on .github/workflows/build.yml:
```diff
@@ -15,7 +15,15 @@ permissions:
+    }}-${{
+      github.event.pull_request.number
+      || github.sha
+    }}
```
Remove whitespace
```suggestion
  group: ${{ github.workflow }}-${{ github.ref_name }}-${{ github.event.pull_request.number || github.sha }}-reusable
```

webknjaz replied:
It was me who asked for wrapped lines FTR.

Also, we decided to drop the reusable suffix. We talked about it with Cam a lot yesterday.

AA-Turner replied:
I don't think wrapped lines meaningfully help here, and because of the interpolation markers it just adds more visual clutter.

Fine with keeping or removing `-reusable`, it wasn't mentioned in the PR so I thought it was unintentional.

vstinner commented on .github/workflows/build.yml:
```diff
@@ -15,7 +15,15 @@ permissions:
 
 concurrency:
-  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}-reusable
+  group: >-
```
Can you please add a comment to explain why we have to put these information into the group? I understand that it's a way to create a global unique identifier.

abstractedfox replied:
Just to clarify, the info from lines 19-26 or just what's on line 18?

webknjaz replied:
The entire thing I imagine.

vstinner replied:
Document how the group string is created.

abstractedfox committed:
Update CI fix & add descriptive comments

Proposed different fix after doing some further testing. This ensures
that concurrency workflows will have a unique name in any situation that
triggers them, and better navigates some quirks in github actions.
The improved comments should make it more visible how the naming
works.

Co-authored-by: Sviatoslav Sydorenko <sviat@redhat.com>
Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

AA-Turner commented on .github/workflows/build.yml:
```diff
@@ -15,7 +15,15 @@ permissions:
+  # the workflow ('Tests'), the branch, and the sha hash
+  #   example: Tests-main-(40 char sha hash)
+  #
+  group: ${{ github.workflow }}-${{ github.ref_name }}-${{ github.head_ref || github.sha }}
```
Remove excess verbosity; simplify (only add 'actor_id')
```suggestion
  # https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#concurrency
  # 'group' must be a key uniquely representing a PR or push event.
  # github.workflow is the workflow name
  # github.actor is the user invoking the workflow
  # github.head_ref is the source branch of the PR or otherwise blank
  # github.run_id is a unique number for the current run
  group: ${{ github.workflow }}-${{ github.actor }}-${{ github.head_ref || github.run_id }}
```

webknjaz replied:
Plz don't. It'll defeat the entire purpose of the cancelation scopes..

AA-Turner replied:
@webknjaz please may you elaborate? The suggestion here is what we use at present, dismbiguated with the user ID. Unsure how this 'defeat[s] the entire purpose'.

hugovk replied:
Shall we use `github.actor` rather than `github.actor_id`? The username is more understandable when needing to read/debug what's going on.

(Theoretically usernames can change, but in practice is unlikely to be of concern.)

AA-Turner replied:
Sure, I've edited the suggestion. 

A

webknjaz replied:
@AA-Turner I forgot to post but we talked this through w/ Kira and concluded that your suggestion was reasonable and I was overthinking it. I originally thought that the fact that it wouldn't produce the same group name for pushes in upstream was undesired but realized later that it's fine.

abstractedfox committed:
Incorporate the triggering actor in concurrency group name

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

hugovk reviewed (approved):
Thanks!

webknjaz reviewed (approved):
LGTM and addresses all of my original concerns.
